src/MongoDB.Driver.Core/Core/Compression/CompressorSource.cs

```diff
@@ -0,0 +1,110 @@
+﻿/* Copyright 2019–present MongoDB Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using MongoDB.Driver.Core.Configuration;
+using MongoDB.Driver.Core.Misc;
+
+namespace MongoDB.Driver.Core.Compression
+{
+    /// <summary>
+    /// Represents a compressor source.
+    /// </summary>
+    public interface ICompressorSource
+    {
+        /// <summary>
+        /// Gets or creates a compressor based on the compressor type.
+        /// </summary>
+        /// <param name="compressorType">The compressor type.</param>
+        /// <returns>The compressor.</returns>
+        ICompressor Get(CompressorType compressorType);
+    }
+
+    internal class CompressorSource : ICompressorSource
+    {
+        private readonly ConcurrentDictionary<CompressorType, ICompressor> _cache;
+        private readonly IReadOnlyList<CompressorConfiguration> _allowedCompressors;
+
+        public CompressorSource(IReadOnlyList<CompressorConfiguration> allowedCompressors)
+        {
+            _cache = new ConcurrentDictionary<CompressorType, ICompressor>();
+            _allowedCompressors = Ensure.IsNotNull(allowedCompressors, nameof(allowedCompressors));
+        }
+
+        public ICompressor Get(CompressorType compressorType)
+        {
+            if (compressorType == CompressorType.Noop)
+            {
+                return null;
+            }
+
+            if (_cache.TryGetValue(compressorType, out var cachedCompressor))
+            {
+                return cachedCompressor;
+            }
+            else
+            {
+                var requestedCompressor = _allowedCompressors.FirstOrDefault(c => c.Type == compressorType);
+                if (requestedCompressor != null)
+                {
+                    return _cache.GetOrAdd(compressorType, CreateCompressor(requestedCompressor));
+                }
+
+                throw new ArgumentException($"The compressor type '{compressorType}' is not allowed.", nameof(compressorType));
+            }
+        }
+
+        public static bool IsCompressorSupported(CompressorType compressorType)
```

Comment:
Put static methods right at the top of the class inside a #region:

    #region static
    ...
    #endregion